index out of range in idpool

```
level=info msg="Conntrack garbage collection statistics" completed=true duration="443.628µs" family=ipv6 maxEntries=262144 numDeleted=0 numKeyFallbacks=0 numLookups=1 numLookupsFailed=0 protocol=non-TCP startTime="2019-01-14 06:00:56.760577245 +0000 UTC m=+244126.543738330" subsys=map-ct
panic: runtime error: index out of range

goroutine 206 [running]:
github.com/cilium/cilium/pkg/idpool.(*idCache).doRemove(0xc421e5c090, 0x8135, 0xc42055ac98)
    /go/src/github.com/cilium/cilium/pkg/idpool/idpool.go:293 +0x151
github.com/cilium/cilium/pkg/idpool.(*idCache).remove(0xc421e5c090, 0x8135, 0xc42180f170)
    /go/src/github.com/cilium/cilium/pkg/idpool/idpool.go:272 +0x35
github.com/cilium/cilium/pkg/idpool.(*IDPool).Remove(0xc42180f170, 0x8135, 0xc42335de00)
    /go/src/github.com/cilium/cilium/pkg/idpool/idpool.go:170 +0x78
github.com/cilium/cilium/pkg/kvstore/allocator.(*cache).start.func1(0xc42182c3c0, 0xc42182c300, 0xc421e27e00, 0xc421d0b450)
    /go/src/github.com/cilium/cilium/pkg/kvstore/allocator/cache.go:186 +0x909
created by github.com/cilium/cilium/pkg/kvstore/allocator.(*cache).start
    /go/src/github.com/cilium/cilium/pkg/kvstore/allocator/cache.go:139 +0x180
```

--- a/pkg/idpool/idpool.go
+++ b/pkg/idpool/idpool.go
@@ -175,10 +175,7 @@ func (p *IDPool) Remove(id ID) bool {
 
 type idCache struct {
 	// ids is a slice of IDs available in this idCache.
-	ids []ID
-
-	// index tracks the position of IDs in the above ids slice.
-	index map[ID]int
+	ids map[ID]struct{}
 
 	// leased is the set of IDs that are leased in this idCache.
 	leased map[ID]struct{}
@@ -191,8 +188,7 @@ func newIDCache(minID ID, maxID ID) *idCache {
 	}
 
 	c := &idCache{
-		ids:    make([]ID, n),
-		index:  make(map[ID]int, n),
+		ids:    make(map[ID]struct{}, n),
 		leased: make(map[ID]struct{}),
 	}
 
@@ -200,8 +196,7 @@ func newIDCache(minID ID, maxID ID) *idCache {
 	seq := random.Perm(n)
 	for i := 0; i < n; i++ {
 		id := ID(seq[i]) + minID
-		c.ids[i] = id
-		c.index[id] = i
+		c.ids[id] = struct{}{}
 	}
 
 	return c
@@ -209,18 +204,12 @@ func newIDCache(minID ID, maxID ID) *idCache {
 
 // allocateID returns a random available ID without leasing it
 func (c *idCache) allocateID() ID {
-	if len(c.ids) == 0 {
-		return NoID
+	for id := range c.ids {
+		delete(c.ids, id)
+		return id
 	}
 
-	// Dequeue the next available ID from the random sequence
-	id := c.ids[0]
-	c.ids = c.ids[1:]
-
-	// Remove from slice of avaiable IDs
-	delete(c.index, id)
-
-	return id
+	return NoID
 }
 
 // leaseAvailableID returns a random available ID.
@@ -265,7 +254,7 @@ func (c *idCache) use(id ID) bool {
 // insert adds the ID into the cache if it is currently unavailable.
 // Returns true if the ID was added to the cache.
 func (c *idCache) insert(id ID) bool {
-	if _, exists := c.index[id]; exists {
+	if _, ok := c.ids[id]; ok {
 		return false
 	}
 
@@ -273,46 +262,19 @@ func (c *idCache) insert(id ID) bool {
 		return false
 	}
 
-	c.ids = append(c.ids, id)
-	c.index[id] = len(c.ids) - 1
-
+	c.ids[id] = struct{}{}
 	return true
 }
 
 // remove removes the ID from the cache.
 // Returns true if the ID was available in the cache.
 func (c *idCache) remove(id ID) bool {
-	removed := c.doRemove(id)
-	// If the id is leased, update it.
 	delete(c.leased, id)
 
-	return removed
-}
-
-func (c *idCache) doRemove(id ID) bool {
-	i, exists := c.index[id]
-	if !exists {
-		return false
+	if _, ok := c.ids[id]; ok {
+		delete(c.ids, id)
+		return true
 	}
 
-	delete(c.index, id)
-
-	N := len(c.ids)
-	if N == 0 {
-		return false
-	}
-
-	tmp := c.ids[N-1]
-	c.ids[i] = tmp
-	if N > 1 {
-		c.ids = c.ids[:N-1]
-	} else {
-		c.ids = c.ids[:0]
-	}
-
-	if id != tmp {
-		c.index[tmp] = i
-	}
-
-	return true
+	return false
 }

--- a/pkg/idpool/idpool.go
+++ b/pkg/idpool/idpool.go
@@ -75,10 +75,6 @@ type IDPool struct {
 
 	// idCache is a cache of IDs backing the pool.
 	idCache *idCache
-
-	// Upon a refresh of the pool, idCache will be pointed to
-	// nextIDCache.
-	nextIDCache *idCache
 }
 
 // NewIDPool returns a new ID pool
@@ -87,28 +83,10 @@ func NewIDPool(minID ID, maxID ID) *IDPool {
 		minID: minID,
 		maxID: maxID,
 	}
-	p.StartRefresh()
-	p.FinishRefresh()
+
+	p.idCache = newIDCache(p.minID, p.maxID)
 
 	return p
-}
-
-// StartRefresh creates a new cache backing the pool.
-// This cache becomes live when FinishRefresh() is called.
-func (p *IDPool) StartRefresh() {
-	c := newIDCache(p.minID, p.maxID)
-
-	p.mutex.Lock()
-	p.nextIDCache = c
-	p.mutex.Unlock()
-}
-
-// FinishRefresh makes the most recent cache created by the pool live.
-func (p *IDPool) FinishRefresh() {
-	p.mutex.Lock()
-	defer p.mutex.Unlock()
-
-	p.idCache = p.nextIDCache
 }
 
 // LeaseAvailableID returns an available ID at random from the pool.
@@ -161,7 +139,7 @@ func (p *IDPool) Insert(id ID) bool {
 	p.mutex.Lock()
 	defer p.mutex.Unlock()
 
-	return p.nextIDCache.insert(id)
+	return p.idCache.insert(id)
 }
 
 // Remove makes an ID unavailable in the pool.
@@ -170,7 +148,7 @@ func (p *IDPool) Remove(id ID) bool {
 	p.mutex.Lock()
 	defer p.mutex.Unlock()
 
-	return p.nextIDCache.remove(id)
+	return p.idCache.remove(id)
 }
 
 type idCache struct {

--- a/pkg/idpool/idpool.go
+++ b/pkg/idpool/idpool.go
@@ -15,7 +15,6 @@
 package idpool
 
 import (
-	"fmt"
 	"math/rand"
 	"strconv"
 	"time"
@@ -92,11 +91,6 @@ func NewIDPool(minID ID, maxID ID) *IDPool {
 	p.FinishRefresh()
 
 	return p
-}
-
-// Dump returns the IDPool as formatted structure for debugging
-func (p *IDPool) Dump() string {
-	return fmt.Sprintf("pool: %+v, cache: %+v", p, p.nextIDCache)
 }
 
 // StartRefresh creates a new cache backing the pool.

--- a/pkg/idpool/idpool_test.go
+++ b/pkg/idpool/idpool_test.go
@@ -18,6 +18,7 @@ package idpool
 
 import (
 	"sort"
+	"sync"
 	"testing"
 
 	"github.com/cilium/cilium/pkg/checker"
@@ -325,4 +326,75 @@ func leaseAllIDs(p *IDPool, minID int, maxID int, c *C) {
 	// Unique ids must have been leased.
 	sort.Ints(actual)
 	c.Assert(actual, checker.DeepEquals, expected)
+}
+
+func (s *IDPoolTestSuite) BenchmarkRemoveIDs(c *C) {
+	minID, maxID := 1, c.N
+	p := NewIDPool(ID(minID), ID(maxID))
+
+	c.ResetTimer()
+	for i := minID; i <= maxID; i++ {
+		c.Assert(p.Remove(ID(i)), Equals, true)
+	}
+}
+
+func (s *IDPoolTestSuite) BenchmarkLeaseIDs(c *C) {
+	minID, maxID := 1, c.N
+	p := NewIDPool(ID(minID), ID(maxID))
+
+	c.ResetTimer()
+	for i := 1; i <= c.N; i++ {
+		id := p.LeaseAvailableID()
+		c.Assert(p.Release(ID(id)), Equals, true)
+	}
+}
+
+func (s *IDPoolTestSuite) BenchmarkUseAndRelease(c *C) {
+	minID, maxID := 1, c.N
+	p := NewIDPool(ID(minID), ID(maxID))
+
+	c.ResetTimer()
+	for i := 1; i <= c.N; i++ {
+		id := p.LeaseAvailableID()
+		c.Assert(p.Use(ID(id)), Equals, true)
+	}
+
+	for i := 1; i <= c.N; i++ {
+		c.Assert(p.Insert(ID(i)), Equals, true)
+	}
+}
+
+func (s *IDPoolTestSuite) TestAllocateID(c *C) {
+	minID, maxID := 1, 6000
+	p := NewIDPool(ID(minID), ID(maxID))
+
+	allocated := make(chan ID, 100)
+	var allocators sync.WaitGroup
+
+	for i := 0; i < 256; i++ {
+		allocators.Add(1)
+		go func() {
+			for i := 1; i <= maxID; i++ {
+				id := p.AllocateID()
+				c.Assert(id, Not(Equals), NoID)
+				allocated <- id
+			}
+			allocators.Done()
+		}()
+	}
+
+	go func() {
+		allocators.Wait()
+		close(allocated)
+	}()
+
+	for {
+		select {
+		case id, ok := <-allocated:
+			if !ok {
+				return
+			}
+			c.Assert(p.Insert(id), Equals, true)
+		}
+	}
 }

--- a/pkg/idpool/idpool_test.go
+++ b/pkg/idpool/idpool_test.go
@@ -99,51 +99,6 @@ func (s *IDPoolTestSuite) TestInsertRemoveIDs(c *C) {
 	c.Assert(actualIDs, checker.DeepEquals, evenIDs)
 }
 
-func (s *IDPoolTestSuite) TestRefresh(c *C) {
-	minID, maxID := 1, 5
-	p := NewIDPool(ID(minID), ID(maxID))
-
-	p.StartRefresh()
-
-	// Lease all ids from the current cache.
-	for i := minID; i <= maxID; i++ {
-		id := p.LeaseAvailableID()
-		c.Assert(id, Not(Equals), NoID)
-	}
-
-	// Delete even-numbered ids from the next cache.
-	for i := minID; i <= maxID; i++ {
-		if i%2 == 0 {
-			c.Assert(p.Remove(ID(i)), Equals, true)
-		}
-	}
-
-	// We should be out of IDs until we finish the refresh.
-	id := p.LeaseAvailableID()
-	c.Assert(id, Equals, NoID)
-
-	p.FinishRefresh()
-
-	// Check we can only lease the remaining
-	// odd-numbered ids from the current cache.
-	expected := make([]int, 0)
-	actual := make([]int, 0)
-	for i := minID; i <= maxID; i++ {
-		if i%2 != 0 {
-			id := p.LeaseAvailableID()
-			c.Assert(id, Not(Equals), NoID)
-			actual = append(actual, int(id))
-			expected = append(expected, int(i))
-		}
-	}
-	// We should be out of IDs.
-	id = p.LeaseAvailableID()
-	c.Assert(id, Equals, NoID)
-
-	sort.Ints(actual)
-	c.Assert(actual, checker.DeepEquals, expected)
-}
-
 func (s *IDPoolTestSuite) TestReleaseID(c *C) {
 	minID, maxID := 1, 5
 	p := NewIDPool(ID(minID), ID(maxID))

--- a/pkg/kvstore/allocator/allocator.go
+++ b/pkg/kvstore/allocator/allocator.go
@@ -42,10 +42,6 @@ const (
 	// performed before failing.
 	maxAllocAttempts = 16
 
-	// allocAttemptsWatermark defines the watermark at what point we will
-	// attempt to clear the cache before re-attempting to allocate
-	allocAttemptsWatermark = 12
-
 	// listTimeout is the time to wait for the initial list operation to
 	// succeed when creating a new allocator
 	listTimeout = 3 * time.Minute
@@ -559,20 +555,6 @@ func (a *Allocator) Allocate(key AllocatorKey) (idpool.ID, bool, error) {
 		}
 
 		kvstore.Trace("Allocation attempt failed", err, logrus.Fields{fieldKey: key, logfields.Attempt: attempt})
-
-		// We have reached a watermark in allocation attempts. The
-		// failure is somewhat persistent. There are multiple reasons
-		// including:
-		// - continued connectivity problem to kvstore
-		// - stale local cache due to backlog in processing of kvstore
-		//   events
-		//
-		// To prevent the stale local ache
-		if attempt == allocAttemptsWatermark {
-			if err := a.mainCache.restart(a); err != nil {
-				log.WithError(err).Warning("Unable to clear and refill allocator cache")
-			}
-		}
 
 		boff.Wait()
 	}

--- a/pkg/kvstore/allocator/cache.go
+++ b/pkg/kvstore/allocator/cache.go
@@ -93,11 +93,6 @@ func (c *cache) getLogger() *logrus.Entry {
 	})
 }
 
-func (c *cache) restart(a *Allocator) error {
-	c.stop()
-	return c.startAndWait(a)
-}
-
 func (c *cache) keyToID(key string, deleteInvalid bool) idpool.ID {
 	if !strings.HasPrefix(key, c.prefix) {
 		invalidKey(key, c.prefix, deleteInvalid)
@@ -132,7 +127,6 @@ func (c *cache) start(a *Allocator) waitChan {
 	c.nextCache = idMap{}
 	c.nextKeyCache = keyMap{}
 	c.mutex.Unlock()
-	a.idPool.StartRefresh()
 
 	c.stopWatchWg.Add(1)
 
@@ -151,7 +145,6 @@ func (c *cache) start(a *Allocator) waitChan {
 					c.cache = c.nextCache
 					c.keyCache = c.nextKeyCache
 					c.mutex.Unlock()
-					a.idPool.FinishRefresh()
 
 					// report that the list operation has
 					// been completed and the allocator is

